File: android/AlpineApp/app/src/main/java/com/alpine/app/data/transport/JsonRpcTransport.kt
package com.alpine.app.data.transport

import com.alpine.app.data.model.QueryRequest
import com.alpine.app.data.model.QueryResponse
import com.alpine.app.data.model.QueryResultsResponse
import com.alpine.app.data.model.QueryStatusResponse
import com.alpine.app.data.rpc.AlpineRpcService

class JsonRpcTransport(private val rpcService: AlpineRpcService) : QueryTransport {

    override suspend fun startQuery(request: QueryRequest): Result<QueryResponse> {
        return try {
            val queryId = rpcService.startQuery(
                queryString = request.queryString,
                groupName = request.groupName,
                autoHaltLimit = request.autoHaltLimit,
                peerDescMax = request.peerDescMax
            )
            Result.success(QueryResponse(queryId))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    override suspend fun getQueryStatus(queryId: Long): Result<QueryStatusResponse> {
        return try {
            val inProgress = rpcService.queryInProgress(queryId)
            val status = rpcService.getQueryStatus(queryId)
            Result.success(status.copy(inProgress = inProgress))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    override suspend fun getQueryResults(queryId: Long): Result<QueryResultsResponse> {
        return try {
            Result.success(rpcService.getQueryResults(queryId))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    override suspend fun cancelQuery(queryId: Long): Result<Unit> {
        return try {
            rpcService.cancelQuery(queryId)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    override fun shutdown() {
        rpcService.shutdown()
    }
}

File: ios/AlpineApp/AlpineApp/Transport/JsonRpcTransport.swift
import Foundation

/// Wraps AlpineRpcService to conform to the QueryTransport protocol,
/// providing query operations over JSON-RPC to an Alpine REST Bridge.
final class JsonRpcTransport: QueryTransport {
    private let rpcService: AlpineRpcService

    init(rpcService: AlpineRpcService) {
        self.rpcService = rpcService
    }

    func startQuery(_ request: QueryRequest) async throws -> QueryResponse {
        let queryId = try await rpcService.startQuery(
            queryString: request.queryString,
            groupName: request.groupName,
            autoHaltLimit: request.autoHaltLimit,
            peerDescMax: request.peerDescMax
        )
        return QueryResponse(queryId: queryId)
    }

    func getQueryStatus(_ queryId: Int64) async throws -> QueryStatusResponse {
        try await rpcService.getQueryStatus(queryId: queryId)
    }

    func getQueryResults(_ queryId: Int64) async throws -> QueryResultsResponse {
        try await rpcService.getQueryResults(queryId: queryId)
    }

    func cancelQuery(_ queryId: Int64) async throws {
        _ = try await rpcService.cancelQuery(queryId: queryId)
    }

    func shutdown() {
        rpcService.shutdown()
    }
}
